roofit/roofitcore/test/TestStatistics/testLikelihoodSerial.cxx

```diff
@@ -0,0 +1,444 @@
+//      std::make_shared<RooFit::TestStatistics::RooUnbinnedL>(pdf, data);
+//      std::make_shared<RooFit::TestStatistics::RooSumL>(pdf, data, RooFit::TestStatistics::GlobalObservables({*w.var("alpha_bkg_obs")}));
+      RooFit::TestStatistics::buildUnbinnedConstrainedLikelihood(pdf, data, RooFit::TestStatistics::GlobalObservables({*w.var("alpha_bkg_obs")}));
+   RooFit::TestStatistics::LikelihoodSerial nll_ts(likelihood, clean_flags/*, nullptr*/);
+
+   nll_ts.evaluate();
+   auto nll1 = nll_ts.getResult();
+
+   EXPECT_EQ(nll0, nll1);
+}
+
+TEST_F(LikelihoodSerialTest, SimUnbinned)
+{
+   // SIMULTANEOUS FIT OF 2 UNBINNED DATASETS
+
+   w.factory("ExtendPdf::egA(Gaussian::gA(x[-10,10],mA[2,-10,10],s[3,0.1,10]),nA[1000])") ;
+   w.factory("ExtendPdf::egB(Gaussian::gB(x,mB[-2,-10,10],s),nB[100])") ;
+   w.factory("SIMUL::model(index[A,B],A=egA,B=egB)") ;
+
+   pdf = w.pdf("model");
+   // Construct dataset from physics pdf
+   data = pdf->generate(RooArgSet(*w.var("x"),*w.cat("index")));
+
+   nll.reset(pdf->createNLL(*data));
+
+   // --------
+
+   auto nll0 = nll->getVal();
+
+   likelihood = RooFit::TestStatistics::buildSimultaneousLikelihood(pdf, data);
+   RooFit::TestStatistics::LikelihoodSerial nll_ts(likelihood, clean_flags/*, nullptr*/);
+
+   nll_ts.evaluate();
+   auto nll1 = nll_ts.getResult();
+
+   EXPECT_EQ(nll0, nll1);
+}
+
+TEST_F(LikelihoodSerialTest, SimUnbinnedNonExtended)
+{
+   // SIMULTANEOUS FIT OF 2 UNBINNED DATASETS
+
+   w.factory("ExtendPdf::egA(Gaussian::gA(x[-10,10],mA[2,-10,10],s[3,0.1,10]),nA[1000])") ;
+   w.factory("ExtendPdf::egB(Gaussian::gB(x,mB[-2,-10,10],s),nB[100])") ;
+   w.factory("SIMUL::model(index[A,B],A=gA,B=gB)") ;
+
+   RooDataSet* dA = w.pdf("gA")->generate(*w.var("x"),1) ;
+   RooDataSet* dB = w.pdf("gB")->generate(*w.var("x"),1) ;
+   w.cat("index")->setLabel("A") ;
+   dA->addColumn(*w.cat("index")) ;
+   w.cat("index")->setLabel("B") ;
+   dB->addColumn(*w.cat("index")) ;
+
+   data = (RooDataSet*) dA->Clone() ;
+   static_cast<RooDataSet*>(data)->append(*dB) ;
+
+   pdf = w.pdf("model");
+
+   nll.reset(pdf->createNLL(*data));
+
+   likelihood = RooFit::TestStatistics::buildSimultaneousLikelihood(pdf, data);
+   RooFit::TestStatistics::LikelihoodSerial nll_ts(likelihood, clean_flags/*, nullptr*/);
+
+   auto nll0 = nll->getVal();
+
+   nll_ts.evaluate();
+   auto nll1 = nll_ts.getResult();
+
+   EXPECT_EQ(nll0, nll1);
+}
+
+
+class LikelihoodSerialSimBinnedConstrainedTest : public LikelihoodSerialTest {
+protected:
+   void SetUp() override {
+      LikelihoodSerialTest::SetUp();
+      // Unbinned pdfs that define template histograms
+
+      w.factory("Gaussian::gA(x[-10,10],-2,3)") ;
+      w.factory("Gaussian::gB(x[-10,10],2,1)") ;
+      w.factory("Uniform::u(x)");
+
+      // Generate template histograms
+
+      RooDataHist* h_sigA = w.pdf("gA")->generateBinned(*w.var("x"),1000) ;
```

Comment:
leak.
Test should be leak free if possible, because then they can be used for leak checking the rest of RooFit.

Reply:
Done, also lots of leaks in RooRealL.cpp, fixed them.